src/ring/middleware/charset.clj

```diff
@@ -0,0 +1,37 @@
+(ns ring.middleware.charset
+  "Middleware for automatically adding a charset to the content-type header in
+  response maps."
+  (:require [ring.util.response :as response]))
+
+(defn- text-based-content-type? [content-type-header]
+  "Is this a text-based content type? True for all text/ types and some
+  application/ types. Non-exhaustive."
+  (or (re-find #"text/" content-type-header)
+      (re-find #"application/(xml|json)" content-type-header)))
+
+(defn- add-charset
+  "Adds a charset to the content-type header in the response map.
+  See: wrap-charset."
+  {:arglists '([response charset])}
+  [resp charset]
+  (if-let [content-type-header (response/get-header resp "Content-Type")]
+    (if-not (re-find #";\s*charset=[^;]*" content-type-header)
+      (if (text-based-content-type? content-type-header)
+        (response/charset resp charset)
+        resp)
+      resp)
+    resp))
+
+(defn wrap-charset
+  "Middleware that adds a charset to the content-type header of the response if
+  one was not set by the handler. Adds the specified charset or defaults to
+  utf-8 if no charset is specified.
+
+  Accepts the following options:
+
+  :charset - the charset to add"
+  {:arglists '([handler] [handler options])}
+  [handler & [opts]]
```

Comment:
While using an option map is usually a good idea for future proofing, because this is a very narrowly scoped middleware function, I think the signature can just be `[handler charset]`.